### PR TITLE
new G1 rules

### PR DESCRIPTION
**Pull request by kirk-microsoft**
This pull requests adds a number of new G1 rules. In the process of adding them, some changes needed to be made to the G1 UL parser

### CODE REVIEW

File: parser/src/main/java/com/microsoft/gctoolkit/parser/G1GCForwardReference.java [hunk 1]
```diff
@@ -774,7 +774,7 @@ public void fullPhase(int integerGroup, String fullGCInternalPhase, double durat
     }
 
     public String toString() {
-        return gcType + ":" + getGCCause() + ":" + getDuration();
+        return getStartTime().toString() + " : " + gcType + " : " + getGCCause() + " : " + getDuration();
```
Comment (dsgrieve) [inline parser/src/main/java/com/microsoft/gctoolkit/parser/G1GCForwardReference.java:777]: Bikeshedding, but I prefer the order to be type, cause, datetimestamp, duration. My reasoning is that it keeps the two time elements together. The 'record' API that I've put into Aggregations follows that order.

File: parser/src/main/java/com/microsoft/gctoolkit/parser/GenerationalHeapParser.java [hunk 2]
```diff
@@ -261,6 +263,9 @@
         parseRules.put(SPLIT_PARNEW_PROMOTION_FAILED_IN_CMS_PHASE, this::parNewPromotionFailedInConcurrentMarkSweepPhase);
         parseRules.put(FULL_SPLIT_BY_CONCURRENT_PHASE, this::fullSplitByConcurrentPhase);
         parseRules.put(CMF_LARGE_BLOCK, this::concurrentModeFailureSplitByLargeBlock);
+        parseRules.put(WEAK_PROCESSING, this::noop);
```
Comment (dsgrieve) [inline parser/src/main/java/com/microsoft/gctoolkit/parser/GenerationalHeapParser.java:266]: Is there a TODO or issue for filling this out?

File: parser/src/main/java/com/microsoft/gctoolkit/parser/unified/UnifiedG1GCPatterns.java [hunk 3]
```diff
@@ -60,11 +60,23 @@
 
     //[73.082s][info ][gc           ] GC(263) Concurrent Cycle
     //[2.179s][info ][gc          ] GC(9) Concurrent Mark Cycle
+//<<<<<<< HEAD
```
Comment (dsgrieve) [inline parser/src/main/java/com/microsoft/gctoolkit/parser/unified/UnifiedG1GCPatterns.java:63]: Looks like you have an unresolved merge conflict. Or did you keep this around for historical reasons?

**Review by dsgrieve — APPROVED** [review]

**Comment by karianna** [comment]
@kirk-microsoft - I think if you rebase and pick up Heinz's fix and the test should pass

File: parser/src/main/java/com/microsoft/gctoolkit/parser/G1GCForwardReference.java (hunk 1 above)
Comment (kirk-microsoft) [reply parser/src/main/java/com/microsoft/gctoolkit/parser/G1GCForwardReference.java:777]: I'm following the convention in the GC logs of time, type, cause, duration

File: parser/src/main/java/com/microsoft/gctoolkit/parser/G1GCForwardReference.java [hunk 4]
```diff
@@ -773,8 +773,13 @@ public void fullPhase(int integerGroup, String fullGCInternalPhase, double durat
...
     }
 
+    /**
+     * Not that this value will be printed but having toString() defined is useful for debugging. The issue is
+     * null values for gctype and gcCause will cause IDE debuggers to NPE. Capture and set to null.
+     * @return String rep of this forward reference.
+     */
     public String toString() {
-        return getStartTime().toString() + " : " + gcType + " : " + getGCCause() + " : " + getDuration();
+        return getStartTime().toString() + " : " + ((gcType == null) ? "null" : gcType.toString()) + " : " + ((getGCCause() == null) ? "null" : getGCCause().toString()) + " : " + getDuration();
```
Comment (dsgrieve) [inline parser/src/main/java/com/microsoft/gctoolkit/parser/G1GCForwardReference.java:782]: String.valueOf(gcType) would work here.

File: parser/src/main/java/com/microsoft/gctoolkit/parser/unified/UnifiedG1GCPatterns.java [hunk 5]
```diff
@@ -60,11 +60,23 @@
...
     //    .... entire set of concurrent records.
     //[73.171s][info ][gc            ] GC(263) Concurrent Cycle 89.437ms
     //[2.179s][info ][gc          ] GC(9) Concurrent Mark Cycle 96.518ms
-    GCParseRule CONCURRENT_CYCLE_START = new GCParseRule("CONCURRENT_CYCLE_START", "Concurrent (?:Mark )?Cycle$");
-    GCParseRule CONCURRENT_CYCLE_END = new GCParseRule("CONCURRENT_CYCLE_END", "Concurrent (?:Mark )?Cycle " + CONCURRENT_TIME);
+    GCParseRule CONCURRENT_CYCLE_START = new GCParseRule("CONCURRENT_CYCLE_START", "Concurrent(?: Mark| Undo)?(?: Cycle)?$");
+    //"[73.171s][info ][gc            ] GC(263) Concurrent Cycle 89.437ms",
+    //"[155.836s][info ][gc          ] GC(2457) Concurrent Undo Cycle 49.351ms",
+    GCParseRule CONCURRENT_CYCLE_END = new GCParseRule("CONCURRENT_CYCLE_END", "Concurrent(?: Undo)? Cycle " + CONCURRENT_TIME);
+//=======
```
Comment (dsgrieve) [inline parser/src/main/java/com/microsoft/gctoolkit/parser/unified/UnifiedG1GCPatterns.java:73]: Lines 73-79 should be removed.

**Review by dsgrieve — APPROVED** [review]

**Pull request merged**